ckanext/s3filestore/views/uploads.py

```diff
@@ -47,11 +49,41 @@ def uploaded_file_redirect(upload_to, filename):
 
     return redirect(url)
 
+def authorized_to_signv4_upload(resource_id):
+    ''' If user is authed to use resource_create then they are
+        authued to use signv4_upload()
+    '''
+    context = {
+        u'model': model,
+        u'session': model.Session,
+        u'user': g.user,
+        u'auth_user_obj': g.userobj
+    }
+
+    resource_dict = toolkit.get_action(u'resource_show')(
+            context, {
+                u'id': resource_id,
+            }
+        )
+
+    return authz.is_authorized('resource_create', context, resource_dict)
+
+
 def signv4_upload():
     aws_secret = ckan_config.get("ckanext.s3filestore.aws_secret_access_key", None)
     region = ckan_config.get("ckanext.s3filestore.region_name", None)
 
     to_sign = str(request.args.get('to_sign')).encode('utf-8')
+    canonical_request = request.args.get('canonical_request')
+
+    # assuming resource path in S3 follows the given structure
+    # '/<BUCKET_NAME>/<FOLDER_NAME>/resources/<RESOURCE_ID>/<FILE_NAME>'
+    resource_id = canonical_request.split('\n')[1].split('/')[4]
```

Comment:
There's one more thing we have to check -- that the `canonical_request` hashes to the same value that we're getting in the `to_sign` string. 

https://docs.aws.amazon.com/IAM/latest/UserGuide/create-signed-request.html#create-string-to-sign